### PR TITLE
Return 400 HTTP response code when the query contains an invalid hex …

### PR DESCRIPTION
**Pull request by softqwewasd**

### CODE REVIEW

File: jaxrs/src/main/java/se/fortnox/reactivewizard/jaxrs/JaxRsRequest.java [hunk 1]
```diff
@@ -108,8 +108,13 @@ public String getQueryParam(String key) {
     public String getQueryParam(String key, String defaultValue) {
         if (queryParameters == null) {
             QueryStringDecoder queryStringDecoder = new QueryStringDecoder(req.uri());
-            queryParameters = queryStringDecoder.parameters();
+            try {
+                queryParameters = queryStringDecoder.parameters();
+            } catch (IllegalArgumentException e) {
+                throw new WebException(HttpResponseStatus.BAD_REQUEST, "invalidhexbyte", "Failed to decode query params due to invalid two digit hexadecimal");
```
Comment (jepp3) [inline jaxrs/src/main/java/se/fortnox/reactivewizard/jaxrs/JaxRsRequest.java:114]: How do we know that the IllegalArgumentException is caused by what we state and assume here?

Comment (softqwewasd) [reply jaxrs/src/main/java/se/fortnox/reactivewizard/jaxrs/JaxRsRequest.java:114]: I made the assumption that IllegalArgumentException is thrown if and only if the hexbyte decoding fails. I could not find other places where an IllegalArgumentException would be thrown for some query param values other than the hexbyte decoding part of the code. That said, I do not think we should pass the exception message to WebException but we can change the message to something more general, like "Failed to decode query parameters" and the "invalidhexbyte" can be changed to "querydecodefailure".

Comment (flowertwig) [reply jaxrs/src/main/java/se/fortnox/reactivewizard/jaxrs/JaxRsRequest.java:114]: Will any of that (except the status code) be exposed in the http response? If so, I don't think we should include it in the WebException, but log the IllegalArgumentException instead.

**Review by flowertwig — CHANGES_REQUESTED** [review]

File: jaxrs/src/main/java/se/fortnox/reactivewizard/jaxrs/JaxRsRequest.java [hunk 2]
```diff
@@ -108,8 +108,14 @@ public String getQueryParam(String key) {
     public String getQueryParam(String key, String defaultValue) {
         if (queryParameters == null) {
             QueryStringDecoder queryStringDecoder = new QueryStringDecoder(req.uri());
-            queryParameters = queryStringDecoder.parameters();
+            try {
+                queryParameters = queryStringDecoder.parameters();
+            } catch (IllegalArgumentException e) {
+                LOG.error(e.getMessage());
```
Comment (flowertwig) [inline jaxrs/src/main/java/se/fortnox/reactivewizard/jaxrs/JaxRsRequest.java:114]: We should have a log message that briefly explains what happened and then include the exception as the last param in LOG.error. Thus we will get the message and stacktrace.

**Review by flowertwig — CHANGES_REQUESTED** [review]

File: jaxrs/src/main/java/se/fortnox/reactivewizard/jaxrs/JaxRsRequest.java [hunk 3]
```diff
@@ -108,8 +108,14 @@ public String getQueryParam(String key) {
     public String getQueryParam(String key, String defaultValue) {
         if (queryParameters == null) {
             QueryStringDecoder queryStringDecoder = new QueryStringDecoder(req.uri());
-            queryParameters = queryStringDecoder.parameters();
+            try {
+                queryParameters = queryStringDecoder.parameters();
+            } catch (IllegalArgumentException e) {
+                LOG.error("Failed to decode HTTP query params for request " + req.method().name() + " " + req.uri(), e);
```
Comment (flowertwig) [inline jaxrs/src/main/java/se/fortnox/reactivewizard/jaxrs/JaxRsRequest.java:114]: Use {} placeholders for the two params.

File: jaxrs/src/test/java/se/fortnox/reactivewizard/jaxrs/JaxRsResourceTest.java [hunk 4]
```diff
@@ -746,6 +746,21 @@ public void shouldAcceptBeanParamInherited() {
...
+        assertBadRequest("/test/accepts?%A");
+        assertBadRequest("/test/accepts?%=");
+        assertBadRequest("/test/accepts?%?");
+        assertBadRequest("/test/accepts?%AG");
+        assertBadRequest("/test/accepts?%A@");
+    }
+
+    private void assertBadRequest(String uri) {
+        MockHttpServerResponse mockHttpServerResponse = get(service, uri);
+        assertThat(mockHttpServerResponse.status()).isEqualTo(HttpResponseStatus.BAD_REQUEST);
```
Comment (flowertwig) [inline jaxrs/src/test/java/se/fortnox/reactivewizard/jaxrs/JaxRsResourceTest.java:761]: Use LoggingVerifier to assert the logging as well.

**Review by flowertwig — CHANGES_REQUESTED** [review]

File: jaxrs/src/test/java/se/fortnox/reactivewizard/jaxrs/JaxRsResourceTest.java (hunk 4 above)
Comment (softqwewasd) [reply jaxrs/src/test/java/se/fortnox/reactivewizard/jaxrs/JaxRsResourceTest.java:768]: :+1:

**Review by flowertwig — APPROVED** [review]

**Pull request merged**